src/WidgetBase.ts

```diff
@@ -231,51 +232,70 @@ export class WidgetBase<P = WidgetProperties, C extends DNode = DNode> extends E
 	}
 
 	/**
-	 * A render decorator that verifies nodes required in
-	 * 'meta' calls in this render,
-	 */
-	@beforeRender()
-	protected verifyRequiredNodes(renderFunc: () => DNode, properties: WidgetProperties, children: DNode[]): () => DNode {
-		return () => {
-			this._requiredNodes.forEach((element, key) => {
-				/* istanbul ignore else: only checking for errors */
-				if (!this._nodeMap.has(key)) {
-					throw new Error(`Required node ${key} not found`);
-				}
-			});
-			this._requiredNodes.clear();
-			const dNodes = renderFunc();
-			this._nodeMap.clear();
-			return dNodes;
-		};
-	}
-
-	/**
 	 * vnode afterCreate callback that calls the onElementCreated lifecycle method.
 	 */
 	private _afterCreateCallback(
-		element: Element,
+		element: HTMLElement,
+		projectionOptions: ProjectionOptions,
+		vnodeSelector: string,
+		properties: VNodeProperties
+	): void {
+		this._nodeHandler.add(element, properties);
+		this.onElementCreated(element, String(properties.key));
+	}
+
+	protected afterRootCreateCallback(
+		element: HTMLElement,
 		projectionOptions: ProjectionOptions,
 		vnodeSelector: string,
 		properties: VNodeProperties
 	): void {
-		this._setNode(element, properties);
+		this._addElementToNodeHandler(element, projectionOptions, properties);
 		this.onElementCreated(element, String(properties.key));
 	}
 
 	/**
 	 * vnode afterUpdate callback that calls the onElementUpdated lifecycle method.
 	 */
 	private _afterUpdateCallback(
-		element: Element,
+		element: HTMLElement,
 		projectionOptions: ProjectionOptions,
 		vnodeSelector: string,
 		properties: VNodeProperties
 	): void {
-		this._setNode(element, properties);
+		this._nodeHandler.add(element, properties);
 		this.onElementUpdated(element, String(properties.key));
 	}
 
+	protected afterRootUpdateCallback(
+		element: HTMLElement,
+		projectionOptions: ProjectionOptions,
+		vnodeSelector: string,
+		properties: VNodeProperties
+	): void {
+		this._addElementToNodeHandler(element, projectionOptions, properties);
+		this.onElementUpdated(element, String(properties.key));
+	}
+
+	private _addElementToNodeHandler(element: HTMLElement, projectionOptions: ProjectionOptions, properties: VNodeProperties) {
+		this._currentRootNode += 1;
+
+		if (this._projectorAttachEvent === undefined) {
+			this._projectorAttachEvent = projectionOptions.nodeEvent.on('attached',
```

Comment:
this isn't really attached or mount (which it previously was), and more after dom update no?